src/stash/src/lib.rs

```diff
@@ -307,3 +442,362 @@ impl From<&str> for StashError {
         }
     }
 }
+
+pub trait Append: Stash {
+    fn append<I: IntoIterator<Item = AppendBatch>>(&self, batches: I) -> Result<(), StashError>;
+}
+
+#[derive(Clone, Debug)]
+pub struct AppendBatch {
+    pub collection_id: Id,
+    pub old_upper: Antichain<Timestamp>,
+    pub new_upper: Antichain<Timestamp>,
+    pub timestamp: Timestamp,
+    pub entries: Vec<((Vec<u8>, Vec<u8>), Timestamp, Diff)>,
+}
+
+impl<K, V> StashCollection<K, V>
+where
+    K: Codec + Ord,
+    V: Codec + Ord,
+{
+    /// Create a new AppendBatch for this collection from its current upper.
+    pub fn make_batch<S: Stash>(&self, stash: &S) -> Result<AppendBatch, StashError> {
+        let old_upper = stash.upper(*self)?;
+        let timestamp: Timestamp = match old_upper.elements() {
+            [ts] => *ts,
+            _ => return Err("cannot determine batch timestamp".into()),
+        };
+        let new_upper = match timestamp.checked_add(1) {
+            Some(ts) => Antichain::from_elem(ts),
+            None => return Err("cannot determine new upper".into()),
+        };
+        Ok(AppendBatch {
+            collection_id: self.id,
+            old_upper,
+            new_upper,
+            timestamp,
+            entries: Vec::new(),
+        })
+    }
+
+    pub fn append_to_batch(&self, batch: &mut AppendBatch, key: &K, value: &V, diff: Diff) {
+        let mut key_buf = vec![];
+        let mut value_buf = vec![];
+        key.encode(&mut key_buf);
+        value.encode(&mut value_buf);
+        batch
+            .entries
+            .push(((key_buf, value_buf), batch.timestamp, diff));
+    }
+}
+
+impl<K, V> From<Id> for StashCollection<K, V> {
+    fn from(id: Id) -> Self {
+        Self {
+            id,
+            _kv: PhantomData,
+        }
+    }
+}
+
+/// A helper struct to prevent mistyping of a [`StashCollection`]'s name and
+/// k,v types.
+pub struct TypedCollection<K, V> {
+    name: &'static str,
+    typ: PhantomData<(K, V)>,
+}
+
+impl<K, V> TypedCollection<K, V> {
+    pub const fn new(name: &'static str) -> Self {
+        Self {
+            name,
+            typ: PhantomData,
+        }
+    }
+}
+
+impl<K, V> TypedCollection<K, V>
+where
+    K: Codec + Ord,
+    V: Codec + Ord,
+{
+    pub fn get(&self, stash: &impl Stash) -> Result<StashCollection<K, V>, StashError> {
+        stash.collection(self.name)
+    }
+
+    pub fn peek_one<S>(&self, stash: &S) -> Result<BTreeMap<K, V>, StashError>
+    where
+        S: Stash,
+        K: Hash,
+    {
+        stash.peek_one(self.get(stash)?)
+    }
+
+    pub fn peek_key_one(&self, stash: &impl Stash, key: &K) -> Result<Option<V>, StashError> {
+        stash.peek_key_one(self.get(stash)?, key)
+    }
```

Comment:
I just want to confirm my understanding about the usage of these methods. Is this correct? 

Collections support multiple values per key and a multiplicity greater than 1. However when we want to make a SQL table like abstraction over collections then the key is used as a primary key and the value is used as the rest of the row. Multiple values per key would violate primary key uniqueness. A multiplicity of 1 indicates that the row exists, but anything higher than 1 doesn't really make sense in that context.

So these methods will validate those constraints over collections when we're using them for a SQL table like abstraction?

Reply:
Correct!